base/task.jl

```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+    done_mask = Bool[]
+    exception = false
+    nremaining::Int = 0
+
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            push!(done_mask, false)
+            nremaining += 1
+        end
+    end
+
+    if nremaining == 0
+        return tasks, Task[]
+    elseif any(done_mask) && (!all || (failfast && exception))
+        return tasks[done_mask], tasks[.~done_mask]
+    end
+
+    chan = Channel{Tuple{Int,Task}}(Inf)
+    waiter_tasks = Dict{Task, Task}()
```

Comment:
Do you think it be sensible to use the index for this, for a vector rather than a map?
```suggestion
    chan = Channel{Int}(Inf)
    waiter_tasks = Vector{Task}(undef, length(tasks))
    # or perhaps, `fill(current_task(), length(tasks))`, and use `current_task()` as a sentinel value instead of using `_unsetindex!`?
```
later then instead of
```
        i, t = take!(chan)
```
using the `tasks` vector
```
        i = take!(chan)
        t = tasks[i]
```
and later instead of
```
for t in remaining_tasks
```
use
```
for t in waiter_tasks[.~done_mask]
end
```

Reply:
I used `current_task()` as a sentinel.